### ISSUE TITLE
Hotgym run fails, linking issue 

### ISSUE THREAD

**Issue by breznak** [htm-community/nupic.cpp#272]
My hotgym run failed, 

> Scanning dependencies of target benchmark_hotgym
[ 98%] Building CXX object src/CMakeFiles/benchmark_hotgym.dir/examples/Hotgym.cpp.o
[100%] Linking CXX executable benchmark_hotgym
/usr/bin/g++-8  -O3 -DNDEBUG   CMakeFiles/benchmark_hotgym.dir/examples/Hotgym.cpp.o  -o benchmark_hotgym -Wl,-rpath,/home/mmm/devel/HTM/htm-community/nupic.cpp/build/scripts/src: libnupic_core.so -lstdc++fs -lpthread -ldl
[100%] Built target benchmark_hotgym
[100%] Built target unit_tests
Install the project...
-- Install configuration: "Release"
-- Installing: /home/mmm/devel/HTM/htm-community/nupic.cpp/build/Release/lib/libnupic_core.a
-- Installing: /home/mmm/devel/HTM/htm-community/nupic.cpp/build/Release/bin/benchmark_hotgym
-- Set runtime path of "/home/mmm/devel/HTM/htm-community/nupic.cpp/build/Release/bin/benchmark_hotgym" to ""
-- Installing: /home/mmm/devel/HTM/htm-community/nupic.cpp/build/Release/lib/libnupic_core.so
-- Installing: /home/mmm/devel/HTM/htm-community/nupic.cpp/build/Release/include/nupic/Version.hpp
-- Installing: /home/mmm/devel/HTM/htm-community/nupic.cpp/build/Release/bin/unit_tests
(pyenv) mmm@mmm-U2442:~/devel/HTM/htm-community/nupic.cpp/build/scripts$ pwd
/home/mmm/devel/HTM/htm-community/nupic.cpp/build/scripts
(pyenv) mmm@mmm-U2442:~/devel/HTM/htm-community/nupic.cpp/build/scripts$ ../Release/bin/benchmark_hotgym
../Release/bin/benchmark_hotgym: error while loading shared libraries: libnupic_core.so: cannot open shared object file: No such file or directory

I noticed from `.travisci` that fix is to 
> export LD_LIBRARY_PATH= 

We should either cmake message that during hotgym build, or put in Readme, or make run seamlessly, if possible?

**Comment by dkeeney** [comment]
Maybe we should static link this so you don't need the library path.
I was looking for a way to be sure that shared linking actually worked in the CI.
Should we link two executables, one with and one without shared linking?

**Comment by breznak** [comment]
> I was looking for a way to be sure that shared linking actually worked in the CI.
Should we link two executables, one with and one without shared linking? 

I don't know, on one hand it's good to have a test that the .so works, on the other it's just another step for first users to run our code. 
So probably the 2 executables, and run .so version under CI.

**Comment by dkeeney** [comment]
> So probably the 2 executables, and run .so version under CI.

Can you put that in or would you prefer that I do it?
Also, it does not have to be hotgym.  We could use unit_tests executable or any other executable.

**Comment by breznak** [comment]
I can do it, but probably no earlier than end-of-week.. 

> Also, it does not have to be hotgym. We could use unit_tests executable or any other executable.

not unittests, those are run even more than hotgym. But you are right, maybe we can just create a dummy executable that prints "Hello HTM world!" and links statically (?)

**Comment by dkeeney** [comment]
There is no hurry on this ... sometime next week even.

Another thought is that a simple "Hello world" might not be a very good test.  It does not insure that all of the expected exported entry points are actually there.  But then... a complete test like that would be difficult and really time-consuming to build.  

The unit_tests executable might actually catch most of that. It was by linking the unit_tests to the shared library that I discovered that the  yaml-cpp module was missing from the library and I had to go back and fix it.  We could build two unit_tests executables, one for static and one for shared. Have one of the CI's run the unit_tests from the shared library. The others from the static version.

Just a thought.  Do what feels sufficient.
